### PR TITLE
Refactor resolve & register for DI

### PR DESCRIPTION
**Pull request by 304NotModified**
- Resolve easier
-  refactor registering, no unregister with `null`, `null` not allowed (update: partly now allowed)
- remove duplicate fields for IValueFormatter
 - Improve IServiceRepository injection into ConfigurationItemFactory
- note: added `TODO DI` comments

### CODE REVIEW

**Comment by 304NotModified** [comment]
@snakefoot please review, thanks!

File: src/NLog/Config/ConfigurationItemFactory.cs [hunk 1]
```diff
@@ -192,8 +194,8 @@ internal LayoutRendererFactory GetLayoutRenderers()
         [Obsolete("Instead use LogFactory.ServiceRepository.ResolveInstance(typeof(IJsonConverter)). Marked obsolete on NLog 5.0")]
         public IJsonConverter JsonConverter
         {
-            get => _serviceResolver.ResolveJsonConverter();
-            set => (_serviceResolver as IServiceRepository)?.RegisterJsonConverter(value);
+            get => _serviceResolver.ResolveService<IJsonConverter>();
```
Comment (snakefoot) [inline src/NLog/Config/ConfigurationItemFactory.cs]: Missing fallback to static instance?

File: src/NLog/LayoutRenderers/LayoutRenderer.cs [hunk 2]
```diff
@@ -304,5 +310,15 @@ public static void Register(string name, Func<LogEventInfo, LoggingConfiguration
             
             ConfigurationItemFactory.Default.GetLayoutRenderers().RegisterFuncLayout(name, layoutRenderer);
         }
+
+        /// <summary>
+        /// Resolve from DI
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        protected T Resolve<T>() where T : class
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/LayoutRenderer.cs]: Think I will skip this, until we have figured out a good way to do this.

File: src/NLog/Layouts/Layout.cs [hunk 3]
```diff
@@ -390,5 +390,16 @@ internal virtual bool TryGetRawValue(LogEventInfo logEvent, out object rawValue)
             rawValue = null;
             return false;
         }
+
+        /// <summary>
+        /// Resolve from DI
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        protected T Resolve<T>() where T : class
```
Comment (snakefoot) [inline src/NLog/Layouts/Layout.cs]: Think I will skip this, until we have figured out a good way to do this.

File: src/NLog/Targets/Target.cs [hunk 4]
```diff
@@ -757,5 +757,15 @@ public static void Register(string name, Type targetType)
             ConfigurationItemFactory.Default.Targets
                 .RegisterDefinition(name, targetType);
         }
+
+        /// <summary>
+        /// Resolve from DI
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        protected T Resolve<T>() where T : class
```
Comment (snakefoot) [inline src/NLog/Targets/Target.cs]: Think I will skip this, until we have figured out a good way to do this.

File: src/NLog/LayoutRenderers/LayoutRenderer.cs [hunk 5]
```diff
@@ -49,12 +49,18 @@ public abstract class LayoutRenderer : ISupportsInitialize, IRenderable, IDispos
...
 
         /// <summary>
         /// Gets the logging configuration this target is part of.
         /// </summary>
         protected LoggingConfiguration LoggingConfiguration { get; private set; }
 
+        /// <summary>
+        /// Value formatter
+        /// </summary>
+        protected IValueFormatter ValueFormatter => _valueFormatter ?? (_valueFormatter = Resolve<IValueFormatter>());
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/LayoutRenderer.cs]: Missing fallback to MessageTemplates.ValueFormatter.Instance

File: src/NLog/Config/ConfigurationItemFactory.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/ConfigurationItemFactory.cs]: It's added already to the container?

File: src/NLog/LayoutRenderers/LayoutRenderer.cs (hunk 5 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/LayoutRenderer.cs]: It's in the container, and you can register `null`

File: src/NLog/Config/ConfigurationItemFactory.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/ConfigurationItemFactory.cs]: So intentional. No need for a default here

File: src/NLog/LayoutRenderers/LayoutRenderer.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/LayoutRenderer.cs]: What isn't good about this?

Comment (304NotModified) [reply src/NLog/LayoutRenderers/LayoutRenderer.cs]: I was think to:

support Ctor injection and the Resolve method for Layouts/target etc. The user could choose. For the NLog components we will use the Resolve Method.

File: src/NLog/Config/ConfigurationItemFactory.cs [hunk 6]
```diff
@@ -202,8 +204,8 @@ public IJsonConverter JsonConverter
         [Obsolete("Instead use LogFactory.ServiceRepository.ResolveInstance(typeof(IValueFormatter)). Marked obsolete on NLog 5.0")]
         public IValueFormatter ValueFormatter
         {
-            get => _serviceResolver.ResolveValueFormatter();
-            set => (_serviceResolver as IServiceRepository)?.RegisterValueFormatter(value);
+            get => _serviceResolver.ResolveService<IValueFormatter>() ?? MessageTemplates.ValueFormatter.Instance;
```
Comment (snakefoot) [inline src/NLog/Config/ConfigurationItemFactory.cs]: If resolver is always right, then you can remove the fallback.

Comment (304NotModified) [reply src/NLog/Config/ConfigurationItemFactory.cs]: good one, fixed

**Pull request merged**